Refactor Steem Account Data Updater: Configuration, Error Handling, and Performance Improvements

BottoStm opened the pull request:
### 📝 Description
This pull request updates the script responsible for updating and maintaining Steem account data and related properties in MongoDB. The changes include moving from environment variables to a configuration file, enhancing error handling, implementing batch processing for Steem API requests, and more structured logging.

### 🔄 Comparison of Initial and Updated Script:

1. **Configuration Management**:
   - **Initial Script**: Used environment variables for `STEEMD_URL` and `MONGODB`.
   - **Updated Script**: Replaced environment variables with a `config.json` file for better configuration management.

2. **Error Handling**:
   - **Initial Script**: Had basic error handling with minimal logging.
   - **Updated Script**: Added robust error handling using the `tenacity` library for retries, detailed logging, and specific error messages to enhance debugging and reliability.

3. **Batch Request Handling**:
   - **Initial Script**: Did not handle batch requests explicitly for the Steem API.
   - **Updated Script**: Implemented JSON-RPC batch requests to optimize the Steem API interactions, reducing the number of API calls and improving performance.

4. **Data Processing and Insertion**:
   - **Initial Script**: Processed and inserted data into MongoDB individually.
   - **Updated Script**: Processes account data in batches and uses `bulk_write` for MongoDB operations, improving performance and ensuring data consistency.

5. **Logging**:
   - **Initial Script**: Used basic `print` statements for logging.
   - **Updated Script**: Replaced `print` statements with the `logging` module for structured and leveled logging, making it easier to track and debug issues.

6. **Scheduler**:
   - **Initial Script**: Used APScheduler for scheduling tasks but did not ensure a graceful shutdown.
   - **Updated Script**: Ensures graceful shutdown of APScheduler, preventing potential data corruption and ensuring tasks complete properly.

### 📈 Improvements:
- **Reliability**: Enhanced error handling and retry mechanisms improve the script's robustness and fault tolerance.
- **Performance**: Batch processing and optimized MongoDB operations significantly improve performance, reducing the load on the API and the database.
- **Maintainability**: Moving to a `config.json` file simplifies configuration management and makes the script easier to maintain and update.
- **Logging**: Structured logging provides better insights into the script's operation, aiding in debugging and monitoring.

### 🚀 How to Test:
1. Update the `config.json` with appropriate values for `STEEMD_URLS` and `MONGODB`.
2. Build and run the Docker container.
3. Monitor the logs to ensure the script runs without errors.
4. Verify the MongoDB collections for correct data insertion.

### ⚙️ Configuration:
Create a `config.json` file in the root directory with the following structure:

```json
{
    "STEEMD_URLS": ["https://api.steemit.com"],
    "MONGODB": "your_mongodb_connection_string"
}

BottoStm committed:
Update README.md

BottoStm committed:
Update README.md

Commit:
Optimize Block Fetching for Faster Blockchain Sync

Modified code to fetch historical blocks in batches of 50 and, once reaching the head block, switch to fetching one block every 3 seconds with retry logic on RPC failure. This approach significantly speeds up the blockchain synchronization process, achieving up to 60% faster sync times in test scenarios.

Changes include:
- Implemented batch fetching of historical blocks, processing up to 50 blocks per batch.
- Added logic to switch to single block fetching mode with a 3-second interval once the head block is reached.
- Introduced retry mechanism on RPC failures to ensure robustness and continuous operation.

These enhancements provide a more efficient and reliable syncing process, allowing for quicker catching up with the latest blocks.

BottoStm committed:
Create Readme.md

BottoStm committed:
Update Readme.md

BottoStm committed:
Update Readme.md

BottoStm committed:
Update Readme.md

BottoStm committed:
Update Readme.md

BottoStm committed:
Merge branch 'master' of https://github.com/BottoStm/steemdb

BottoStm committed:
Batching Operations

Batching Operations and change Dockerfile

BottoStm committed:
Update history.py

BottoStm committed:
Update Dockerfile

BottoStm committed:
Update Dockerfile

BottoStm committed:
Update Dockerfile

BottoStm committed:
Update history.py

Import re module: Added the import statement for the re module.
Fixed logger.error call: Corrected the logger.error call to ensure proper formatting of the error message.

BottoStm committed:
Update history.py

BottoStm committed:
Update history.py

BottoStm committed:
Update history.py

Fixed the TypeError: Updated the get_batch_account_details function to use next(rpc.nodes) to get the current node from the cycle instead of using indexing.

BottoStm committed:
Added Readme

BottoStm committed:
Added Readme

BottoStm committed:
Update Readme.md

ety001 commented on docker/history/Readme.md:
````diff
@@ -0,0 +1,64 @@
+
+3. Run the Docker container:
+   ```sh
+   docker run -d --name steem-history steem-history
````
It need add `config.json` file in params.
like this
```
docker run -d --name steem-history -v ./config.json:/src/config.json steem-history
```

 But I suggest to also add environment configuration way.

BottoStm replied:
Volume flag is not required as config.json is also  copied inside the container , Script now can take steemd url and mongo as environment variable too ,  

```sh
   docker run -d --name steem-history -e STEEMD_URLS="http://10.10.100.12:8080" -e MONGODB="mongodb://10.10.100.30:27017" steem-history
   ```

BottoStm committed:
Update Dockerfile

BottoStm committed:
Update Readme.md

BottoStm committed:
Update history script to support configuration via both config.json and environment variables

- Modified Dockerfile to install necessary dependencies and Python packages.
- Updated history.py to load configuration from config.json if present, otherwise fallback to environment variables.
- Added detailed README.md to document the setup, configuration, and usage of the Steem account data updater.

BottoStm committed:
Update Readme.md

BottoStm committed:
Update Readme.md

ety001 commented on docker/history/history.py:
```diff
@@ -1,44 +1,66 @@
+        config = json.load(config_file)
+    steemd_urls = config.get('STEEMD_URLS', ['https://api.steemit.com'])
+    mongodb_url = config.get('MONGODB')
+except FileNotFoundError:
```
I think this part will not execute in docker style. Because you have copyed a config.json into the docker image.

I suggest we could rename the `config.json` to `config.json.example` .

BottoStm replied:
yes , users can decide Not to copy config.json , that way they can pass Environment Variables Directly , ! Thank you for the isnights , i will do the now , its just rename ,and update to Readme too

BottoStm committed:
Renamed config.jsot to config.jsob.example

Renamed config.json to config.jsob.example   . Developers can Decide to use config.json or directly pass env variables

BottoStm committed:
Update Readme.md
